####Inputs####
##set working directory
setwd('~/todd/Data/Homer_heatmaps')

##set directories for homerTag directories
##Homer tag directories generated from ChIP-seq/ATAC-seq mapped sequencing data using the following command:
##makeTagDirectory <destination directory> <ATAQseq/ChIPseq data> -format sam -unique
tag.dir = '~/chris/Homer_Tags/'
tag.list =c("ES_ATAC","ES_H3K27ac","ES_H3K4me1","ES_Nanog","ES_Oct4","ES_Sox2",
            "TS_ATAC","TS_Cdx2","TS_Elf5","TS_Eomes","TS_H3K27ac","TS_H3K4me1",
            "ES_H3K27me3","TS_H3K27me3","ES_H3K9me3",
            "TKO_DNase_WT","TKO_DNase_TKO","TKO_K27_WT","TKO_K27_TKO",
            "Med1","Med12","Nipbl","p300","Smc1","Smc3",
            "ESC_ATAC_2i","ESC_ATAC_serum")
 
##Set bed file for ROIs as generated by Putting_together_beds_into_ROI
bed = '~/todd/Data/Output/ROI_coord.txt'

##Set path to homer
homer = '~/Homer/bin/'

##import functions
source('/data/Blizard-BrancoLab/Chris/Todd_et_al/Functions/annotatePeaks.R')


####Script####
###Annotate Peaks function from Homer Package
temp.bed.file<-tempfile()
bed<-read.delim(bed)[,1:6]
write.table(bed,file=temp.bed.file,quote=F,sep='\t',col.names=F,row.names=F)
for(i in 1:length(tag.list)){
  full.tag.dir<-paste(tag.dir,tag.list[i],sep="")
  tag.hist<-annotatepeaks(bed.file = temp.bed.file,tag=full.tag.dir,ref.genome = ' mm10 ',path.to.homer = homer)
  write.table(tag.hist,paste(tag.list[i],"_hist.txt",sep=""),sep="\t",quote=F,row.names = F,col.names = T)
}
unlink(temp.bed.file)




